### PR TITLE
[#1432] Implement  pipeline stage

### PR DESCRIPTION
**Pull request by shibasisp**
# Description

Closes #{1432}.

Implemented `$unset` aggregation pipeline. Currently this works with:
- [] single field
- [] multiple fields (array);
- [] dot notation support for both.

## Readiness checklist

* [ ] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [ ] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [ ] I ensured that PR title is good enough for the changelog.
* [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
* [ ] I marked all done items in this checklist.

### CODE REVIEW

File: integration/aggregate_documents_compat_test.go [hunk 1]
```diff
@@ -1562,3 +1562,47 @@ func TestAggregateCompatProject(t *testing.T) {
...
 	testAggregateStagesCompat(t, testCases)
 }
+
+func TestAggregateCompatUnset(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]aggregateStagesCompatTestCase{
+		"InvalidType": {
+			pipeline: bson.A{
+				bson.D{{"$unset", "invalid"}},
```
Comment (rumyantseva) [inline integration/aggregate_documents_compat_test.go:1572]: Why is is called `InvalidType`? This example looks like a valid `$unset` to me, where we unset a field called `invalid`. If it's about an invalid type, then maybe we can use something like  `bson.D{{"$unset",  false}},` (then the type `bool` is really invalid. 

Let's also add one more test where wrong types are set in an array. For example, something like `{$unset: ["v", int32(42), false]}`

File: integration/aggregate_documents_compat_test.go [hunk 2]
```diff
@@ -1562,3 +1562,47 @@ func TestAggregateCompatProject(t *testing.T) {
...
+			pipeline: bson.A{
+				bson.D{{"$unset", "v"}},
+			},
+		},
+		"UnsetID": {
+			pipeline: bson.A{
+				bson.D{{"$unset", "_id"}},
+			},
+		},
+		"Unset2Fields": {
```
Comment (rumyantseva) [inline integration/aggregate_documents_compat_test.go:1586]: I suggest also adding a test case where we unset both `_id` and `v` fields. So, in the results, we will see some empty docs for some providers.

File: integration/aggregate_documents_compat_test.go [hunk 3]
```diff
@@ -1562,3 +1562,47 @@ func TestAggregateCompatProject(t *testing.T) {
...
+				bson.D{{"$unset", "invalid"}},
+			},
+			resultType: emptyResult,
+		},
+		"Unset1Field": {
+			pipeline: bson.A{
+				bson.D{{"$unset", "v"}},
+			},
+		},
+		"UnsetID": {
```
Comment (rumyantseva) [inline integration/aggregate_documents_compat_test.go:1581]: This test case is a bit tricky: because we unset the field `_id`, we need to specify how we want to sort the results. Otherwise, our compat tests might fail because target and compat systems will return results that are sorted in a different way (or where sorting might vary). To do this, I suggest sorting first and then doing unsets. 

So the pipeline will be something like this:
```go
pipeline: bson.A{
	bson.D{{"$sort", bson.D{{"_id", 1}}}},
	bson.D{{"$unset", "_id"}},
},
```

File: internal/handlers/common/aggregations/stages/unset.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		defer fieldIter.Close()
+		for {
+			_, field, err := fieldIter.Next()
+			if err != nil {
+				if errors.Is(err, iterator.ErrIteratorDone) {
+					break
+				}
+				return nil, lazyerrors.Error(err)
+			}
+			fieldStr := field.(string)
```
Comment (rumyantseva) [inline internal/handlers/common/aggregations/stages/unset.go:66]: If fields are set something like `[1, 2, 3]`, it will panic. I think we need a more accurate check here by returning an error if type is not valid.

**Review by rumyantseva — CHANGES_REQUESTED** [review]
Thank you very much for contributing! 

I see the PR is a draft, but the review was requested, so I leave some comments.

I like the implementation, but I suggest making some additions in tests (see my comments), so we can reveal some problems or errors that need special handling.

Could you also run `task lint` please? We have some specific rules to format code, it helps us keep the code consistent.

**Comment by AlekSi** [comment]
@shibasisp Any news? Anything we could help with?

**Comment by shibasisp** [comment]
Apologies for the delay. @AlekSi I will be addressing the comments by this weekend.

**Review by AlekSi — CHANGES_REQUESTED** [review]
It does not compile

**Review by rumyantseva — DISMISSED** [review]
LGTM. Thanks for contributing!

File: internal/handlers/common/aggregations/stages/unset.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+	return &unset{
+		field: fieldsToUnset,
+	}, nil
+}
+
+// Process implements Stage interface.
+func (u *unset) Process(_ context.Context,
+	iter types.DocumentsIterator, closer *iterator.MultiCloser,
+) (types.DocumentsIterator, error) {
```
Comment (AlekSi) [inline internal/handlers/common/aggregations/stages/unset.go:107]:
```suggestion
func (u *unset) Process(_ context.Context, iter types.DocumentsIterator, closer *iterator.MultiCloser) (types.DocumentsIterator, error) { //nolint:lll // for readability
```

We don't wrap lines like that

File: integration/aggregate_documents_compat_test.go [hunk 6]
```diff
@@ -145,7 +145,6 @@ func testAggregateStagesCompatWithProviders(t *testing.T, providers shareddata.P
 					}
 				})
 			}
-
```
Comment (AlekSi) [inline integration/aggregate_documents_compat_test.go:148]: Let's keep that newline

**Review by AlekSi — APPROVED** [review]

File: internal/handlers/common/aggregations/stages/unset.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	field *types.Document
+}
+
+// newUnset validates unset document and creates a new $unset stage.
+func newUnset(stage *types.Document) (aggregations.Stage, error) {
+	fields, err := stage.Get("$unset")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	fieldsToUnset := must.NotFail(types.NewDocument())
```
Comment (noisersup) [inline internal/handlers/common/aggregations/stages/unset.go:45]:
```suggestion
        // fieldsToUnset contains keys with `false` values. They are used to specify projection exclusion later.
	fieldsToUnset := must.NotFail(types.NewDocument())
```

File: internal/handlers/common/aggregations/stages/unset.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+	switch fields := fields.(type) {
+	case *types.Array:
+		if fields.Len() == 0 {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrStageUnsetNoPath,
+				"$unset specification must be a string or an array with at least one field",
+				"$unset (stage)",
+			)
+		}
```
Comment (noisersup) [inline internal/handlers/common/aggregations/stages/unset.go:55]: It would be nice to add a couple of testcases to cover those errors if possible.

**Review by noisersup — APPROVED** [review]
Thanks for your contribution! The logic here looks good to me. I would only suggest to add a comment as it's not clear at the first glance that `ProjectionDocument` would use exclusion and not inclusion for sure.

Also we might add 3 test cases for $unset field to make a code coverage slightly better:
- empty array
- empty string
- array with empty string

File: internal/handlers/common/aggregations/stages/unset.go (hunk 8 above)
Comment (rumyantseva) [reply internal/handlers/common/aggregations/stages/unset.go:55]: I agree with @noisersup, for this line we can add a test case with an empty array as `$unset` param.

File: internal/handlers/common/aggregations/stages/unset.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+					"$unset specification must be a string or an array containing only string values",
+					"$unset (stage)",
+				)
+			}
+
+			fieldsToUnset.Set(fieldStr, false)
+		}
+	case string:
+		if fields == "" {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
```
Comment (rumyantseva) [inline internal/handlers/common/aggregations/stages/unset.go:83]: And for this line, we can add a test case with an empty string as `$unset` param.

**Review by rumyantseva — COMMENTED** [review]
Let's add the last couple of tests to cover error codes fully!

**Review by rumyantseva — COMMENTED** [review]
Let's add the last couple of tests to cover error codes fully!

**Review by chilagrow — APPROVED** [review]
Great, thanks for your amazing contribution :hugs:

**Review by noisersup — APPROVED** [review]
